File: macos-app/Moku/Formats/CbzParser.swift
import Foundation
import ZIPFoundation

/// Parses CBZ (Comic Book ZIP) archives — extracts sorted image pages.
enum CbzParser {

    struct CbzMetadata {
        let title: String
        let author: String
        let pageCount: Int
    }

    // MARK: - Cache

    private nonisolated(unsafe) static var archiveCache: [String: Archive] = [:]
    private nonisolated(unsafe) static var sortedEntries: [String: [Entry]] = [:]

    static func clearCache(_ key: String) {
        archiveCache.removeValue(forKey: key)
        sortedEntries.removeValue(forKey: key)
    }

    // MARK: - Metadata

    static func extractMetadata(from url: URL) -> CbzMetadata {
        let title = url.deletingPathExtension().lastPathComponent
        guard let archive = try? Archive(url: url, accessMode: .read) else {
            return CbzMetadata(title: title, author: "Unknown", pageCount: 0)
        }
        let images = imageEntries(from: archive)
        return CbzMetadata(title: title, author: "Unknown", pageCount: images.count)
    }

    // MARK: - Chapters (pages)

    static func getChapters(filePath: String, booksDir: URL) throws -> [ChapterInfo] {
        let entries = try loadOrCache(filePath: filePath, booksDir: booksDir)
        return entries.enumerated().map { (i, entry) in
            let name = (entry.path as NSString).lastPathComponent
            return ChapterInfo(index: i, title: "Page \(i + 1) — \(name)")
        }
    }

    static func getPageCount(filePath: String, booksDir: URL) throws -> Int {
        try loadOrCache(filePath: filePath, booksDir: booksDir).count
    }

    /// Get image data for a specific page.
    static func getPageImage(filePath: String, pageIndex: Int, booksDir: URL) throws -> Data? {
        let entries = try loadOrCache(filePath: filePath, booksDir: booksDir)
        guard pageIndex >= 0, pageIndex < entries.count else { return nil }

        let url = booksDir.appendingPathComponent(filePath)
        guard let archive = try? Archive(url: url, accessMode: .read) else { return nil }

        let entry = entries[pageIndex]
        var imageData = Data()
        _ = try archive.extract(entry) { chunk in
            imageData.append(chunk)
        }
        return imageData
    }

    /// Extract cover (first image) data.
    static func extractCover(from url: URL) -> Data? {
        guard let archive = try? Archive(url: url, accessMode: .read) else { return nil }
        let images = imageEntries(from: archive)
        guard let first = images.first else { return nil }
        var data = Data()
        _ = try? archive.extract(first) { chunk in data.append(chunk) }
        return data
    }

    // MARK: - Private

    private static func loadOrCache(filePath: String, booksDir: URL) throws -> [Entry] {
        if let cached = sortedEntries[filePath] { return cached }
        let url = booksDir.appendingPathComponent(filePath)
        guard let archive = try? Archive(url: url, accessMode: .read) else {
            throw NSError(domain: "CbzParser", code: 1, userInfo: [NSLocalizedDescriptionKey: "Cannot open archive"])
        }
        let entries = imageEntries(from: archive)
        sortedEntries[filePath] = entries
        return entries
    }

    private static func imageEntries(from archive: Archive) -> [Entry] {
        let imageExtensions: Set<String> = ["jpg", "jpeg", "png", "gif", "webp", "bmp", "tiff"]

        return archive
            .filter { entry in
                guard entry.type == .file else { return false }
                let path = entry.path
                // Skip macOS metadata
                if path.contains("__MACOSX") || path.contains(".DS_Store") { return false }
                let name = (path as NSString).lastPathComponent
                if name.hasPrefix(".") { return false }
                let ext = (name as NSString).pathExtension.lowercased()
                return imageExtensions.contains(ext)
            }
            .sorted { naturalSort($0.path, $1.path) }
    }

    /// Natural sort: "page2" < "page10"
    private static func naturalSort(_ a: String, _ b: String) -> Bool {
        a.compare(b, options: [.numeric, .caseInsensitive]) == .orderedAscending
    }
}

File: app/lib/core/formats/cbz/cbz_parser.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:archive/archive.dart';
import 'package:path/path.dart' as p;

import '../../services/book_service.dart';

/// Metadata extracted from a CBZ (Comic Book ZIP) file.
class CbzMetadata {
  final String title;
  final String author;
  final int pageCount;

  const CbzMetadata({
    required this.title,
    required this.author,
    required this.pageCount,
  });
}

/// Cached CBZ data — sorted image file names + archive bytes.
class _CachedCbz {
  final List<ArchiveFile> images;
  final Archive archive;

  const _CachedCbz({required this.images, required this.archive});
}

/// Parser for CBZ (Comic Book ZIP) files.
///
/// CBZ files are ZIP archives containing sequentially-named image files.
/// Each image is one "page" of the comic.
class CbzParser {
  static final _cache = <String, _CachedCbz>{};

  static const _imageExtensions = {
    '.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp', '.tiff',
  };

  /// Extract metadata from a CBZ file.
  static CbzMetadata extractMetadata(Uint8List bytes, String filePath) {
    final archive = ZipDecoder().decodeBytes(bytes);
    final images = _getSortedImages(archive);

    return CbzMetadata(
      title: p.basenameWithoutExtension(filePath),
      author: 'Unknown',
      pageCount: images.length,
    );
  }

  /// Extract the first image as cover art.
  static Uint8List? extractCover(Uint8List bytes) {
    final archive = ZipDecoder().decodeBytes(bytes);
    final images = _getSortedImages(archive);
    if (images.isEmpty) return null;
    final content = images.first.content as List<int>;
    return Uint8List.fromList(content);
  }

  /// Get chapter (page) list for reader UI.
  static Future<List<ChapterInfo>> getChapters(String filePath) async {
    final cached = await _loadCbz(filePath);
    return cached.images.asMap().entries.map((e) {
      return ChapterInfo(index: e.key, title: 'Page ${e.key + 1}');
    }).toList();
  }

  /// Get a single page image as bytes.
  static Future<Uint8List> getPageImage(
      String filePath, int pageIndex) async {
    final cached = await _loadCbz(filePath);
    if (pageIndex < 0 || pageIndex >= cached.images.length) {
      return Uint8List(0);
    }
    final content = cached.images[pageIndex].content as List<int>;
    return Uint8List.fromList(content);
  }

  /// Get the media type of a page image.
  static String getMediaType(String fileName) {
    final ext = p.extension(fileName).toLowerCase();
    return switch (ext) {
      '.png' => 'image/png',
      '.gif' => 'image/gif',
      '.webp' => 'image/webp',
      '.bmp' => 'image/bmp',
      _ => 'image/jpeg',
    };
  }

  /// Total page count for a cached CBZ.
  static Future<int> getPageCount(String filePath) async {
    final cached = await _loadCbz(filePath);
    return cached.images.length;
  }

  static void clearCache(String filePath) => _cache.remove(filePath);

  // ── Private ───────────────────────────────────────────────────────────────

  static Future<_CachedCbz> _loadCbz(String filePath) async {
    if (_cache.containsKey(filePath)) return _cache[filePath]!;

    final file = File(filePath);
    final bytes = await file.readAsBytes();
    final archive = ZipDecoder().decodeBytes(bytes);
    final images = _getSortedImages(archive);
    final cached = _CachedCbz(images: images, archive: archive);
    _cache[filePath] = cached;
    return cached;
  }

  /// Get image files from archive, sorted by filename (natural order).
  static List<ArchiveFile> _getSortedImages(Archive archive) {
    return archive.files
        .where((f) =>
            !f.isFile ? false : _isImageFile(f.name) && !_isHidden(f.name))
        .toList()
      ..sort((a, b) => _naturalCompare(a.name, b.name));
  }

  static bool _isImageFile(String name) {
    final ext = p.extension(name).toLowerCase();
    return _imageExtensions.contains(ext);
  }

  static bool _isHidden(String name) {
    return p.basename(name).startsWith('.') ||
        name.contains('__MACOSX') ||
        name.contains('.DS_Store');
  }

  /// Natural sort comparison — handles "page2" vs "page10" correctly.
  static int _naturalCompare(String a, String b) {
    final regExp = RegExp(r'(\d+)');
    final aName = p.basename(a);
    final bName = p.basename(b);

    final aMatches = regExp.allMatches(aName).toList();
    final bMatches = regExp.allMatches(bName).toList();

    if (aMatches.isNotEmpty && bMatches.isNotEmpty) {
      // Compare by first number found in filename
      final aNum = int.parse(aMatches.first.group(0)!);
      final bNum = int.parse(bMatches.first.group(0)!);
      if (aNum != bNum) return aNum.compareTo(bNum);
    }

    return aName.compareTo(bName);
  }
}
